workspace/update-injected-packages/src/DirPatcher.ts

```diff
@@ -0,0 +1,208 @@
+
+export type Value = Symlink | File | Dir
+export type InodeMap = Record<string, Value>
+
+export interface DiffItemBase {
+  path: string
+  oldValue?: Value
+  newValue?: Value
+}
+
+export interface AddedItem extends DiffItemBase {
+  path: string
+  oldValue?: undefined
+  newValue: Value
+}
+
+export interface RemovedItem extends DiffItemBase {
+  path: string
+  oldValue: Value
+  newValue?: undefined
+}
+
+export interface ModifiedItem extends DiffItemBase {
+  path: string
+  oldValue: Value
+  newValue: Value
+}
+
+export interface DirDiff {
+  added: AddedItem[]
+  removed: RemovedItem[]
+  modified: ModifiedItem[]
+}
+
+// length comparison should place every directory before the files it contains because
+// a directory path is always shorter than any file path it contains
+const comparePaths = (a: string, b: string): number => (a.split(/\\|\//).length - b.split(/\\|\//).length) || b.localeCompare(a)
+
+/**
+ * Get the difference between 2 files tree.
+ *
+ * The arrays in the resulting object are sorted in such a way that every directory paths are placed before
+ * the files it contains. This way, it would allow optimization for operations upon this diff.
+ * Note that when performing removal of removed files according to this diff, the `removed` array should be reversed first.
+ */
+export function diffDir (oldIndex: InodeMap, newIndex: InodeMap): DirDiff {
+  const oldPaths = Object.keys(oldIndex).sort(comparePaths)
+  const newPaths = Object.keys(newIndex).sort(comparePaths)
+
+  const removed: RemovedItem[] = oldPaths
+    .filter(path => !(path in newIndex))
+    .map(path => ({ path, oldValue: oldIndex[path] }))
+
+  const added: AddedItem[] = newPaths
+    .filter(path => !(path in oldIndex))
+    .map(path => ({ path, newValue: newIndex[path] }))
+
+  const modified: ModifiedItem[] = oldPaths
+    .filter(path => path in newIndex && oldIndex[path] !== newIndex[path])
+    .map(path => ({ path, oldValue: oldIndex[path], newValue: newIndex[path] }))
+
+  return { added, removed, modified }
+}
+
+/**
+ * Apply a diff on a directory.
+ *
+ * The {@link optimizedDirDiff} is assumed to be already optimized (i.e. `removed` is already reversed).
+ */
+export async function applyDiff (optimizedDirDiff: DirDiff, sourceDir: string, targetDir: string): Promise<void> {
+  async function addRecursive (sourcePath: string, targetPath: string, value: Value): Promise<void> {
+    const makeParent = () => fs.mkdirSync(path.dirname(targetPath), { recursive: true })
+    if (value === DIR) {
+      await fs.promises.mkdir(targetPath, { recursive: true })
+    } else if (typeof value === 'string') {
+      makeParent()
+      const symlinkTarget = value
+      if (process.platform === 'win32') {
+        const symlinkRealTarget = path.resolve(sourceDir, symlinkTarget)
+        await fs.promises.symlink(symlinkRealTarget, targetPath, 'junction')
+      } else {
+        await fs.promises.symlink(symlinkTarget, targetPath)
+      }
```

Comment:
we have a library for creating symlinks. symlink-dir

Reply:
https://github.com/pnpm/pnpm/pull/9100/commits/dd0d6a03ea941e554dd785b4405b93bd906c8331